### PR TITLE
fix(graphql-relational-transformer): nullability enforcement for references relational fields

### PR DESCRIPTION
**Pull request by atierian**
### 1. Validate relational fields are nullable
Relational fields **MUST** be nullable. These changes validate this rule.

The following is valid because the fields `Primary.relatedOne`, `Primary.relatedMany`, `RelatedOne.primary` and `RelatedMany.primary` are all nullable / non-required fields.
```gql
type Primary @model { 
  id: ID!
  relatedOne: RelatedOne @hasOne(references: "primaryId")
  relatedMany: [RelatedMany] @hasMany(references: "primaryId")
}

type RelatedOne @model { 
  id: ID!
  primaryId: ID
  primary: Primary @belongsTo(references: "primaryId")
}

type RelatedMany @model { 
  id: ID!
  primaryId: ID
  primary: Primary @belongsTo(references: "primaryId")
}
```

Any of these changes are invalid:
```diff
-  relatedOne: RelatedOne @hasOne(references: "primaryId")
+  relatedOne: RelatedOne! @hasOne(references: "primaryId")
```

```diff
-  relatedMany: [RelatedMany] @hasMany(references: "primaryId")
+  relatedMany: [RelatedMany]! @hasMany(references: "primaryId")
+  relatedMany: [RelatedMany!] @hasMany(references: "primaryId")
+  relatedMany: [RelatedMany!]! @hasMany(references: "primaryId")
```

```diff
-  primary: Primary @belongsTo(references: "primaryId")
+  primary: Primary! @belongsTo(references: "primaryId")
```

### 2. Validate consistent nullability of Related model's reference fields
Reference fields **MUST** have consistent nullability. These changes fix a bug in the existing validation implementation and improve the error message.

##### CDK / CloudFormation Parameters Changed
N/A

#### Issue #, if available
N/A

#### Description of how you validated changes

Unit tests

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] ~Relevant documentation is changed or added (and PR referenced)~
- [ ] ~New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies~
- [ ] ~Any CDK or CloudFormation parameter changes are called out explicitly~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-graphql-relational-transformer/src/__tests__/amplify-graphql-has-many-references-transformer.test.ts [hunk 1]
```diff
@@ -185,7 +185,7 @@ test('fails if used as a has one relationship', () => {
     type Member @model {
       id: ID!
       teamID: String
-      team: Team @belongsTo(fields: ["teamID"])
+      team: Team @belongsTo(references: ["teamID"])
```
Comment (atierian) [inline packages/amplify-graphql-relational-transformer/src/__tests__/amplify-graphql-has-many-references-transformer.test.ts:188]: unrelated fix in test schema

**Review by palpatim — DISMISSED** [review]

**Review by AaronZyLee — APPROVED** [review]

**Review by palpatim — APPROVED** [review]

**Review by palpatim — APPROVED** [review]

**Review by AaronZyLee — APPROVED** [review]
